crates/bevy_ecs/src/query/state.rs

```diff
@@ -1392,53 +1392,92 @@ impl<D: QueryData, F: QueryFilter> QueryState<D, F> {
             // SAFETY: We only access table data that has been registered in `self.archetype_component_access`.
             let tables = unsafe { &world.storages().tables };
             let archetypes = world.archetypes();
-            for storage_id in &self.matched_storage_ids {
-                if D::IS_DENSE && F::IS_DENSE {
-                    let table_id = storage_id.table_id;
-                    let table = &tables[table_id];
-                    if table.is_empty() {
-                        continue;
+            let mut batch_queue = vec![];
```

Comment:
I'm not the biggest fan of needing to allocate a growable vec repeatedly. Could you try using `arrayvec`, with a reasonable max queue size (128 seems more than reasonable).

Reply:
benchmark using arrayvec

![image](https://github.com/bevyengine/bevy/assets/45868716/65f19028-215a-40c7-9f1b-6a4387d1798e)